sdb/commands/ptype.py

```diff
@@ -0,0 +1,92 @@
+#
+# Copyright 2020 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import argparse
+from typing import Iterable
+
+import drgn
+import sdb
+
+
+class PType(sdb.Command):
+    """
+    Print the type passed as an argument.
+
+    EXAMPLES
+        Resolve spa_t typedef:
+
+            sdb> ptype spa_t
+            typedef struct spa spa_t
+
+        Print the members of the spa and vdev structures:
+
+            sdb> ptype spa vdev
+            struct spa {
+                    char spa_name[256];
+                    char *spa_comment;
+                    ...
+            }
+            struct vdev {
+                    uint64_t vdev_id;
+                    uint64_t vdev_guid;
+                    ...
+            }
+
+        Print enums and unions:
+
+            sdb> ptype zfs_case v_t thread_union
+            enum zfs_case {
+                    ZFS_CASE_SENSITIVE = 0,
+                    ZFS_CASE_INSENSITIVE = 1,
+                    ZFS_CASE_MIXED = 2,
+            }
+            typedef union {
+                    iv_t e;
+                    uint8_t b[8];
+            } v_t
+            union thread_union {
+                    struct task_struct task;
+                    unsigned long stack[2048];
+            }
+    """
+
+    names = ["ptype"]
+
+    @classmethod
+    def _init_parser(cls, name: str) -> argparse.ArgumentParser:
+        parser = super()._init_parser(name)
+        parser.add_argument("type", nargs="*", metavar="<type>")
+        return parser
+
+    def _call(self, objs: Iterable[drgn.Object]) -> None:
+        for tp_name in self.args.type:
+            found = False
+
+            for prefix in ["struct ", "enum ", "union ", ""]:
```

Comment:
[nit] we might want to check the exact name they said first (i.e. list `""` first)

Reply:
I just realized that I should have created a comment about this but I'd like to know your opinion.

There are cases in C codebases that someone may do this:
```
typedef union GCObject GCObject;
```
[example from `module/lua/lobject.h` in the ZFS repo (actually Lua repo)]

This is kind of annoying because using `ptype` if we have `""` before all the other strings in that list we get:
```
sdb> ptype GCObject
typedef union GCObject GCObject
```

If we have the code as I have it though, typedef's are the last thing that we check:
```
sdb> ptype GCObject
union GCObject {
	GCheader gch;
	union TString ts;
	union Udata u;
	union Closure cl;
	struct Table h;
	struct Proto p;
	struct UpVal uv;
	struct lua_State th;
}
```

It is an interesting issue which would have been nice if we supported things like `ptype union GCObject` and `ptype GCObject`.

Reply:
Comment added.